### ISSUE TITLE
Running training on multicore server with one core being occupied results in slow performance

### ISSUE THREAD

**Issue by tvollbrecht** [dmlc/xgboost#8729]
We are running xgboost on a multicore server where resources are shared with other processes. We noticed a performance degradation when there is load on the server (the runtime increases from 10 seconds to several minutes) 

We could reproduce the issue with the following simplified setup

gettingstarted.py
```
from xgboost import XGBClassifier

from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split

data = load_digits()
X_train, X_test, y_train, y_test = train_test_split(
    data["data"], data["target"], test_size=0.2

)
bst = XGBClassifier(random_state=42)
bst.fit(X_train, y_train)
preds = bst.predict(X_test)
```

We are testing it running within a docker container generated by the following Dockerfile

```
FROM python:3.8.15-slim
COPY gettingstarted.py .
RUN pip install xgboost==1.7.1 scikit-learn==1.0.2

ENTRYPOINT ["python3", "gettingstarted.py"]
```

Server: `gcp e2-standard-8` (8 core, /proc/cpuinfo reports intel xeon 2.2ghz)
Host system: `Debian 11, 5.10.0-19-cloud-amd64 #1 SMP Debian 5.10.149-2 (2022-10-21) x86_64 GNU/Linux`

Runtime without any load on the server
`time docker run <generateddockerimage>`: `~2.3 sec`

To generate load on one core we are using the command `stress -c 1`.
When running the docker container after load is generated on one core the runtime increases to `7min 15sec`

The same effect can be observed when using the latest versions of xgboost (1.7.3) and scikit-learn (1.2.1)
Runtime: 2.3 sec without load, 7min 22 sec with load one one core

Changing the n_jobs parameter when instantiating the XGBoostClassifier causes the effect to be visible if more cpus are occupied by other processes:
`bst = XGBClassifier(random_state=42, n_jobs=7)`
Runtime without load: 2.3 sec
Runtime with load on one core (`stress -c 1`): 2.4 sec
Runtime with load on two cores (`stress -c 2`): 6min 20 sec

Additional tests:
The effect can be observed when running the code without docker on ubuntu (22.04lts, python 3.10 - runtimes 1.6sec without load, 1m25s with load on one core)
The effect cannot be observed on MacOS (13.1) (python 3.8.15 - runtimes 2.3sec without load / 2.8sec with load on one core)

**Comment by trivialfis** [comment]
XGBoost relies on GCC openmp runtime to schedule threads,which in tun relies on the system scheduler. The system scheduler can be configured, for instance docker has its own scheduler and there are a couple of schedulers implemented in the Linux kernel for users to choose.

I don't think the xgboost library can handle system level scheduling, users need to have there own resource management framework like slurm to handle system resources.

**Comment by tvollbrecht** [comment]
Thank you for the hint regarding OpenMP. I was able to figure out a mitigation for my problem when running it on a server where some cores are already occupied.

For future readers:
I am setting the environment variable OMP_WAIT_POLICY=passive. Without this flag in the gcc version OpenMP uses spin locks waiting for the resource. As the cpu is occupied by another process, this spin lock stay active until the other process yields or is removed by the system scheduler. My assumption is that this causes a chain of spin locks waiting for each other.
The clang implementation uses pthread_yield -> this causes the effect to not be visible on MacOS.
Downside of using this flag (in gcc environment): When running on an idle server the code ran ~30% slower. In my case I will leave the flag activated until we deploy a resource management system hiding occupied cores.
